### PR TITLE
support rust-style tuple

### PR DESCRIPTION
**Pull request by yjh0502**
A goal of this PR is enable rust-style, type-safe tuple decoding, like

```rust
let data: &[u8] = &[2, 104, 101, 108, 108, 111, 0, 1, 119, 111, 114, 108, 100, 0];
let (v1, v2): (String, Vec<u8>) = Tuple::decode(data).expect("Failed to decode tuple");
```

I think at the end there will be two types of tuple API
 - encode/decode arbiraty type of tuple (`Tuple` type in #37). It is similar with `serde_json::Value`, which can be used to decode/encode arbitrary shape of tuples.
 - strong-type tuple. It it similar with `serde_json::from_str`, which decodes the data with expected types, and returns error if type mismatches.

As @bluejekyll is working for on first type of API, I wrote some code for second type of API.

Here's list of works before merging this PR
 - [ ] clean-up names. I borrowed name `Tuple` and `Single` from #37, so it should be changed to something else.
 - [x] floating-point types.
 - [x] integer encoding.
 - [x] support for tuple encoding.
 - [ ] nested tuple.
 - [ ] clean-up constants.

Here's some future works
 - Supporting versionstamp.
 - ...

### CODE REVIEW

File: foundationdb/src/tuple.rs [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+					Ok(out)
+				}
+            }
+        )+
+    }
+}
+
+tuple_impls! {
+    1 => (0 T0)
```
Comment (bluejekyll) [inline foundationdb/src/tuple.rs]: Very cool!

File: foundationdb/src/tuple.rs [hunk 2]
```diff
@@ -0,0 +1,364 @@
+use byteorder::{self, ByteOrder};
+use std;
+use std::io::Write;
+
+#[derive(Debug, Clone, Copy)]
+pub enum TupleError {
```
Comment (bluejekyll) [inline foundationdb/src/tuple.rs]: Should we make this a Failure?

File: foundationdb/src/tuple.rs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+            return Err(TupleError::InvalidData);
+        }
+        Ok(val)
+    }
+}
+
+impl Single for bool {
+    fn encode<W: Write>(&self, mut w: W) -> std::io::Result<()> {
+        if *self {
+            w.write_all(&[0x27])
```
Comment (bluejekyll) [inline foundationdb/src/tuple.rs]: We should define an enum for these codes, you can copy that from my initial tuple impl...

Comment (bluejekyll) [reply foundationdb/src/tuple.rs]: See the SingleType from that with it’s associated to_code impl.

**Comment by bluejekyll** [comment]
@yjh0502 I really like the work here. Do you want to land this, and then we can work on this together?

**Comment by yjh0502** [comment]
@bluejekyll I added an patches which is required for #39 .

**Comment by bluejekyll** [comment]
Shall I merge this in as is?

**Comment by bluejekyll** [comment]
> clean-up names. I borrowed name Tuple and Single from #37, so it should be changed to something else.

we can choose whatever here. Probably stick with Tuple, and then for the inner unit, just Item, maybe, to be consistent with Iterator, etc.

**Pull request merged**